posts/inside-rust/2020-11-02-Using-rustc_cranelift_codegen.md

```diff
@@ -0,0 +1,129 @@
+---
+layout: post
+title: "Using rustc_codegen_cranelift for debug builds"
+author: Joshua Nelson
+team: The Compiler Team <https://www.rust-lang.org/governance/teams/compiler>
+---
+
+## What is `rustc_codegen_cranelift`?
+
+[`rustc_codegen_cranelift`], or just `cg_clif` for short, is a new experimental
+codegen backend for the Rust compiler. The existing backend is LLVM, which very
+good at producing fast, highly optimized code, but is not very good at
+compiling code quickly. `cg_clif`, which uses the [Cranelift] project, would
+provide a fast backend which greatly improves compile times, at the cost of
+very few optimizations. This is a great fit for debug builds, and the hope is
+that `cg_clif` will eventually be the default backend in debug mode.
+
+## What is the progress of using `rustc_codegen_cranelift` for debug builds?
+
+There has been an [Major Change Proposal][MCP] open for some time for making
+`cg_clif` part of the main Rust repository. Recently, [the MCP was
+accepted][compiler-team#270] and the compiler team [merged][#77975]
+`rustc_cranelift_codegen` [into the main Rust git repository][#77975].
+`cg_clif` still is not distributed with `rustup`, but this means you can now
+build it from source in-tree!
+
+## How do I use `rustc_codegen_cranelift`?
```

Comment:
Should this also mention how to build it from the main repo? That is a lot quicker.

Reply:
Yeah, that's what I meant by 'I feel silly'. I'll move building in-tree to a separate section and use the `prepare.sh` for the rest of the article.

Reply:
@bjorn3 is there a way to build cg_clif without also running the tests?

Reply:
actually, test.sh didn't work either:
```
[BUILD] sysroot
error: could not execute process `/home/joshua/rustc_codegen_cranelift/build_sysroot/../target/debug/cg_clif_build_sysroot -vV` (never executed)

Caused by:
  No such file or directory (os error 2)

real	0m0.043s
user	0m0.036s
sys	0m0.010s
```

Reply:
Not at the moment. I can add something.

Reply:
>  test.sh didn't work either:

I think this is because I have `CARGO_TARGET_DIR` set; I unset it in the parent but I guess running a bash shell set it again.

Reply:
I will unset it in `build_sysroot.sh`.

Reply:
Unsetting `CARGO_TARGET_DIR` didn't help. I think there's some cache invalidation issue, maybe?

Reply:
bjorn3/rustc_codegen_cranelift@324e63de289e249481445a399fcbcad62b7ab71d unsets `CARGO_TARGET_DIR` where necessary.

Reply:
Just saw your previous comment. Did you build in debug mode? (not passing `--release` to `test.sh`)

Reply:
Well, either way it works with `build --release && build_sysroot.sh` directly, so probably no need to worry about it.